sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3344,43 +3345,6 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
         } else {
           v2Write
         }
-
-      case u: UpdateTable if !u.skipSchemaResolution && u.resolved =>
```

Comment:
This resolution was substantially different compared to what we do in normal writes or in data sources that actually support row-level operations.

Reply:
Can we highlight the differences? I thought we are already doing by-name assignments here.

Reply:
One notable difference is not using `AssertNotNull` and relying on the same framework we have for V2 tables. In particular, we check the assignment mode and whether in and out attributes are compatible.

The main contribution of this PR is `AssignmentUtils` that aligns assignments with table attributes. It is a prerequisite for rewriting UPDATEs. Right now, we only rewrite DELETEs. It also handles casts and char/varchar types.

Reply:
It's more common in other databases that, if you write values to a non-nullable column, runtime null check is applied instead of rejecting the INSERT because the value is nullable.

Reply:
I agree it varies from database to database but why do we have different validation behavior for regular inserts? I am a bit worried we would have the same problem like with V1 sources when the validation was different in SQL and DataFrame writes. It felt to me V2 commands should behave in the same way.

It is not a blocker for this PR but I'd be curious to understand our long-term plan. Do we plan to switch to runtime checks for regular inserts too?

Reply:
Also, the current logic does not seem to cover nullability in inner checks.

Reply:
I think runtime null check is the right direction if we want to unify but we can collect feedback from more people.

Reply:
I don't have a strong opinion on runtime error or analysis error, but I agree that we need to unify the behaviors.

If I remember correctly, DB2 has a Query Transformation and Semantics layer. The null check is done there instead of run time, but I don't know how this is done in other databases.

Reply:
I see value in both depending on the use case. What about making it configurable? If we just switch to runtime checks everywhere, it will be a substantial behavior change. We can add a new SQL property and default to the existing behavior of throwing an exception during the analysis phase.

By the way, I don't target 3.4 in this PR so we will have time to build a proper runtime checking framework. I think that would be a substantial effort as we need to cover inner fields. There is no logic for that at the moment, if I am not mistaken.

I do think consistency would be important. UPDATE and INSERT should behave in the same way.

Reply:
The only reason I would consider adding a config is to avoid an implicit behavior change as we currently check field compatibility for INSERTS during analysis. What is the community policy on such things? Is it okay to just transition to runtime checks in 3.5?

Reply:
I just don't feel that analysis error to runtime error is a breaking behavior change. Failure scenarios are still failure scenarios, though the timing is different. I'm open to have a config for this if others think it is necessary.

Reply:
I don't feel strongly about this. Looks like we agree about unifying the behavior, which I think is important. I'll wait a bit for others to comment. If no other input, I'll migrate INSERTs to runtime checks then.

Reply:
Okay, since there is no more feedback, I'll follow up with a PR to migrate INSERT validation to runtime checks. After that, I'll come back to this PR. That way, we will have consistent behavior in all operations.

Thanks for the discussion, @cloud-fan @dongjoon-hyun @huaxingao @viirya!

Reply:
I've submitted #40655 to migrate to runtime checks.